src/main/drivers/dshot_bitbang.c

```diff
@@ -0,0 +1,710 @@
+    }
+}
+
+static void bbWrite(uint8_t motorIndex, float value)
+{
+    bbWriteInt(motorIndex, value);
+}
+
+static void bbUpdateComplete(void)
+{
+    // If there is a dshot command loaded up, time it correctly with motor update
+
+    if (!dshotCommandQueueEmpty()) {
+        if (!dshotCommandOutputIsEnabled(bbDevice.count)) {
+            return;
+        }
+    }
+
+#ifdef USE_DSHOT_TELEMETRY
+    for (int i = 0; i < usedMotorPorts; i++) {
+        bbPort_t *bbPort = &bbPorts[i];
+
+        if (useDshotTelemetry) {
+            if (bbPort->direction == DSHOT_BITBANG_DIRECTION_INPUT) {
+                bbPort->inputActive = false;
+                bbSwitchToOutput(bbPort);
+            }
+
+
+            bbDMA_Cmd(bbPort, ENABLE);
+        }
+    }
+#endif
+
+    for (int i = 0; i < usedMotorPacers; i++) {
+        bbPacer_t *bbPacer = &bbPacers[i];
+        bbTIM_DMACmd(bbPacer->tim, bbPacer->dmaSources, ENABLE);
+    }
+}
+
+static bool bbEnableMotors(void)
+{
+    for (int i = 0; i < motorCount; i++) {
+        IOConfigGPIO(bbMotors[i].io, bbMotors[i].iocfg);
+    }
+    return true;
+}
+
+static void bbDisableMotors(void)
+{
+    return;
+}
+
+static void bbShutdown(void)
+{
+    return;
+}
+
+static bool bbIsMotorEnabled(uint8_t index)
+{
+    return bbMotors[index].enabled;
+}
+
+static const motorDevConfig_t *lastMotorConfig;
+
+static void bbPostInit()
+{
+    bbFindPacerTimer();
+
+    for (int motorIndex = 0; motorIndex < MAX_SUPPORTED_MOTORS && motorIndex < motorCount; motorIndex++) {
+
+        if (!bbMotorConfig(bbMotors[motorIndex].io, motorIndex, lastMotorConfig->motorPwmProtocol, bbMotors[motorIndex].output)) {
+            return NULL;
+        }
+
+
+        bbMotors[motorIndex].enabled = true;
+
+        // Fill in motors structure for 4way access (XXX Should be refactored)
+
+        motors[motorIndex].enabled = true;
+    }
+}
+
+static motorVTable_t bbVTable = {
+    .postInit = bbPostInit,
+    .enable = bbEnableMotors,
+    .disable = bbDisableMotors,
+    .isMotorEnabled = bbIsMotorEnabled,
+    .updateStart = bbUpdateStart,
+    .write = bbWrite,
+    .writeInt = bbWriteInt,
+    .updateComplete = bbUpdateComplete,
+    .convertExternalToMotor = dshotConvertFromExternal,
+    .convertMotorToExternal = dshotConvertToExternal,
+    .shutdown = bbShutdown,
+};
+
+
+motorDevice_t *bbshotDevInit(const motorDevConfig_t *motorConfig, uint16_t idlePulse, uint8_t count, bool useUnsyncedPwm)
+{
+    UNUSED(idlePulse);
+    UNUSED(useUnsyncedPwm);
+
+    dbgPinInit();
+    dbgPinLo(0);
+    dbgPinLo(1);
+
+    lastMotorConfig = motorConfig;
+    bbDevice.vTable = bbVTable;
+    motorCount = count;
+
+#ifdef USE_DSHOT_TELEMETRY
+    useDshotTelemetry = motorConfig->useDshotTelemetry;
+#endif
+
+    for (int motorIndex = 0; motorIndex < MAX_SUPPORTED_MOTORS && motorIndex < motorCount; motorIndex++) {
+        const timerHardware_t *timerHardware = timerGetByTag(motorConfig->ioTags[motorIndex]);
```

Comment:
Is this timer actually available for other uses when bitbanging Dshot is used? If not, then I this should use `timerAllocate(motorConfig->ioTags[motorIndex], OWNER_DSHOT_BITBANG, motorIndex)` to mark the timer as used.

Reply:
These timers are available for other uses. They are only accessed here to check whether the signal should be inverted.

Reply:
Isn't pin inversion hardwired into Timer PWM? There should be no inversion necessary for bitbang ...